File: C++/039. Combination Sum.cpp
class Solution {
private: 
    void SearchSum(vector<vector<int>> &res, vector<int>& candidates, vector<int>& sol, int target, int index)
    {
        for(int j=index;candidates[j]<=target && j < candidates.size();j++)
        {
            
            if (candidates[j]<=target)
            {
                sol.push_back(candidates[j]);
                if(candidates[j]==target)
                {
                    res.push_back(sol);
                    sol.pop_back();
                    return;
                }
                else
                {
                    SearchSum(res, candidates, sol, target-candidates[j], j);
                    sol.pop_back();
                }
            }
            else{break;}
        }
        return;
    }

public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> sol;
        vector<vector<int>> res;
        int size = candidates.size();
        if (size<=0){return res;}
        sort(candidates.begin(),candidates.end());
        SearchSum(res, candidates, sol, target, 0);
        return res;
    }
};

File: python/039. Combination Sum.py
class Solution(object):
    def dfs(self, candidates, target, idx, cur, res):
        if target == 0: 
            res.append(copy.copy(cur))
            return
        if idx >= len(candidates) or candidates[idx] > target: return
        for i in range(idx, len(candidates)):
            if candidates[i] <= target:
                cur.append(candidates[i])
                self.dfs(candidates, target-candidates[i], i, cur, res)
                cur.pop()
            else:
                break
        return
    
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if target == 0: return []
        candidates.sort()
        cur = []
        res = []
        self.dfs(candidates, target, 0, cur, res)
        return res